File: src/lib/emit.ts
import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import { harnessAdapters } from "../adapters/index.js";
import {
  canonicalMcpServers,
  type HarnessName,
  type HooksSource,
  type ManifestComponentPaths,
  type PluginMetadata,
  PORTABLE_EVENTS,
  type PortableEvent,
  type PortableHooks,
  pluginMetadataSchema,
} from "../domain/harness.js";

const BOOTSTRAP_COMMAND_MARKER = "hooks/cheese-bootstrap.sh";

function filterPortableEvents(source: HooksSource): PortableHooks {
  const portable: PortableHooks = {};
  const portableSet: ReadonlySet<string> = new Set(PORTABLE_EVENTS);

  for (const [event, entries] of Object.entries(source)) {
    if (entries === undefined) continue;
    if (portableSet.has(event)) {
      portable[event as PortableEvent] = entries;
    } else {
      console.warn(`[cheese-flow] skipping non-portable hook event: ${event}`);
    }
  }

  return portable;
}

function filterBootstrapEntries(portable: PortableHooks): PortableHooks {
  const result: PortableHooks = {};
  for (const [event, entries] of Object.entries(portable)) {
    if (entries === undefined) continue;
    const kept = entries.filter(
      (entry) => !entry.command.includes(BOOTSTRAP_COMMAND_MARKER),
    );
    if (kept.length > 0) {
      result[event as PortableEvent] = kept;
    }
  }
  return result;
}

export async function emitPluginManifest(
  harness: HarnessName,
  metadata: PluginMetadata,
  outputRoot: string,
  componentPaths: ManifestComponentPaths = {},
): Promise<string> {
  const parsed = pluginMetadataSchema.safeParse(metadata);
  if (!parsed.success) {
    throw new Error(
      `Plugin metadata validation failed: ${parsed.error.message}`,
    );
  }

  const adapter = harnessAdapters[harness];
  const manifest = adapter.buildManifest(parsed.data, componentPaths);
  const manifestDir = path.join(outputRoot, adapter.manifestDir);

  await mkdir(manifestDir, { recursive: true });
  const manifestPath = path.join(manifestDir, "plugin.json");
  await writeFile(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`);
  return manifestPath;
}

export async function emitMcpConfig(
  harness: HarnessName,
  outputRoot: string,
): Promise<string> {
  await mkdir(outputRoot, { recursive: true });

  const config: Record<string, unknown> = {
    mcpServers: canonicalMcpServers,
  };

  const adapter = harnessAdapters[harness];
  const outputPath = path.join(outputRoot, adapter.mcpFileName);
  await writeFile(outputPath, `${JSON.stringify(config, null, 2)}\n`);
  return outputPath;
}

export async function emitHooks(
  harness: HarnessName,
  source: HooksSource,
  outputRoot: string,
): Promise<false | string> {
  const adapter = harnessAdapters[harness];
  let portable = filterPortableEvents(source);
  if (!adapter.capabilities.bootstrapHook) {
    portable = filterBootstrapEntries(portable);
  }
  const payload = adapter.buildHookConfig(portable);

  if (payload === null) {
    console.info(
      `[cheese-flow] hooks not supported in ${harness} target; skipping`,
    );
    return false;
  }

  await mkdir(outputRoot, { recursive: true });
  const outputPath = path.join(outputRoot, "hooks.json");
  await writeFile(outputPath, `${JSON.stringify(payload, null, 2)}\n`, "utf8");
  return outputPath;
}

File: python/cheese_flow/lib/emit.py
"""Port of `src/lib/emit.ts` — write plugin manifest, MCP config, hooks files."""

from __future__ import annotations

import json
import sys
from pathlib import Path

from cheese_flow.adapters import HARNESS_ADAPTERS
from cheese_flow.lib.harness import (
    CANONICAL_MCP_SERVERS,
    PORTABLE_EVENTS,
    HarnessName,
    HooksSource,
    ManifestComponentPaths,
    PluginMetadata,
    PortableEvent,
    PortableHooks,
)

_BOOTSTRAP_COMMAND_MARKER = "hooks/cheese-bootstrap.sh"


def _filter_portable_events(source: HooksSource) -> PortableHooks:
    portable: PortableHooks = {}
    portable_set: frozenset[str] = frozenset(PORTABLE_EVENTS)
    for event, entries in source.items():
        if entries is None:
            continue
        if event in portable_set:
            portable[event] = entries  # type: ignore[index]
        else:
            print(
                f"[cheese-flow] skipping non-portable hook event: {event}",
                file=sys.stderr,
            )
    return portable


def _filter_bootstrap_entries(portable: PortableHooks) -> PortableHooks:
    result: PortableHooks = {}
    for event, entries in portable.items():
        if entries is None:
            continue
        kept = [
            entry
            for entry in entries
            if _BOOTSTRAP_COMMAND_MARKER not in entry["command"]
        ]
        if len(kept) > 0:
            event_key: PortableEvent = event  # type: ignore[assignment]
            result[event_key] = kept
    return result


def _dump_json(payload: object) -> str:
    """Match Node ``JSON.stringify(value, null, 2)`` formatting (with a trailing newline)."""
    return json.dumps(payload, indent=2, ensure_ascii=False) + "\n"


def emit_plugin_manifest(
    harness: HarnessName,
    metadata: PluginMetadata,
    output_root: str | Path,
    component_paths: ManifestComponentPaths | None = None,
) -> Path:
    component_paths = component_paths or {}
    adapter = HARNESS_ADAPTERS[harness]
    manifest = adapter.buildManifest(metadata, component_paths)
    manifest_dir = Path(output_root) / adapter.manifestDir
    manifest_dir.mkdir(parents=True, exist_ok=True)
    manifest_path = manifest_dir / "plugin.json"
    manifest_path.write_text(_dump_json(manifest), encoding="utf-8")
    return manifest_path


def emit_mcp_config(harness: HarnessName, output_root: str | Path) -> Path:
    adapter = HARNESS_ADAPTERS[harness]
    Path(output_root).mkdir(parents=True, exist_ok=True)
    config = {"mcpServers": CANONICAL_MCP_SERVERS}
    output_path = Path(output_root) / adapter.mcpFileName
    output_path.write_text(_dump_json(config), encoding="utf-8")
    return output_path


def emit_hooks(
    harness: HarnessName,
    source: HooksSource,
    output_root: str | Path,
) -> Path | None:
    adapter = HARNESS_ADAPTERS[harness]
    portable = _filter_portable_events(source)
    if not adapter.capabilities.bootstrapHook:
        portable = _filter_bootstrap_entries(portable)
    payload = adapter.buildHookConfig(portable)

    if payload is None:
        # Mirror TS ``console.info`` — informational output goes to stdout so
        # ``cheese compile`` log scrapers see the message in the same stream.
        print(f"[cheese-flow] hooks not supported in {harness} target; skipping")
        return None

    Path(output_root).mkdir(parents=True, exist_ok=True)
    output_path = Path(output_root) / "hooks.json"
    output_path.write_text(_dump_json(payload), encoding="utf-8")
    return output_path
